python/ray/train/tests/test_tune.py

```diff
@@ -299,6 +313,47 @@ def train_func():
     assert len(trial_dfs[0]["training_iteration"]) == 4
 
 
+@pytest.mark.parametrize("num_gpus_per_worker", [0.5, 1, 2])
+def test_tune_torch_get_device_gpu(ray_2_node_4_gpu, num_gpus_per_worker):
+    from ray import tune
+    from ray.tune.tuner import Tuner, TuneConfig
+
+    num_samples = 2
+
+    @patch("torch.cuda.is_available", lambda: True)
+    def train_func():
+        train.report(device_id=train.torch.get_device().index)
+
+    trainer = TorchTrainer(
+        train_func,
+        torch_config=TorchConfig(backend="gloo"),
+        scaling_config={
+            "num_workers": 2,
+            "use_gpu": True,
+            "resources_per_worker": {"GPU": num_gpus_per_worker},
+        },
+    )
+
+    tuner = Tuner(
+        trainer,
+        param_space={
+            "train_loop_config": {
+                "dummy": tune.choice([32, 64, 128]),
+            }
+        },
+        tune_config=TuneConfig(
+            num_samples=num_samples,
+        ),
+    )
+    analysis = tuner.fit()._experiment_analysis
+    trial_dfs = list(analysis.trial_dataframes.values())
+    device_ids = [trial_df["device_id"].tolist() for trial_df in trial_dfs]
+
+    assert len(device_ids) == num_samples
+    for i in range(num_samples):
+        assert device_ids[i][0] == 0
```

Comment:
Hmm what exactly is this test supposed to be testing and how reliable is it? If we have 2 trials, each with 1 gpu (2*0.5) then the placement of the trials can be on either node, and we will get different results each time?

Reply:
thanks, right! the results will be still the same, because in this setup, the `device_id` is `almost` equal to the local rank. That being said, the device_id is always being `0` (because ray.tune separates the environment).